test/integration/plugin_test.go

```diff
@@ -0,0 +1,528 @@
+package integration
+
+import (
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/kubernetes-incubator/kube-aws/core/root"
+	"github.com/kubernetes-incubator/kube-aws/core/root/config"
+	"github.com/kubernetes-incubator/kube-aws/model"
+	"github.com/kubernetes-incubator/kube-aws/plugin"
+	"github.com/kubernetes-incubator/kube-aws/test/helper"
+)
+
+func TestPlugin(t *testing.T) {
+	kubeAwsSettings := newKubeAwsSettingsFromEnv(t)
+
+	s3URI, s3URIExists := os.LookupEnv("KUBE_AWS_S3_DIR_URI")
+
+	if !s3URIExists || s3URI == "" {
+		s3URI = "s3://examplebucket/exampledir"
+		t.Logf(`Falling back s3URI to a stub value "%s" for tests of validating stack templates. No assets will actually be uploaded to S3`, s3URI)
+	}
+
+	mainClusterYaml := kubeAwsSettings.mainClusterYaml()
+	minimalValidConfigYaml := mainClusterYaml + `
+availabilityZone: us-west-1c
+`
+	validCases := []struct {
+		context       string
+		clusterYaml   string
+		plugins       []helper.TestPlugin
+		assertConfig  []ConfigTester
+		assertCluster []ClusterTester
+	}{
+		{
+			context: "WithAddons",
+			clusterYaml: minimalValidConfigYaml + `
+kubeAwsPlugins:
+  myPlugin:
+    enabled: true
+    queue:
+      name: baz1
+    oidc:
+      issuer:
+        url: "https://login.example.com/"
+
+worker:
+  nodePools:
+  - name: pool1
+    kubeAwsPlugins:
+      myPlugin:
+        enabled: true
+        queue:
+          name: baz2
+`,
+			plugins: []helper.TestPlugin{
+				helper.TestPlugin{
+					Name: "my-plugin",
+					Files: map[string]string{
+						"assets/controller/baz.txt": "controller-baz",
+						"assets/etcd/baz.txt":       "etcd-baz",
+						"assets/worker/baz.txt":     "worker-baz",
+					},
+					Yaml: `
+metadata:
+  name: my-plugin
+  version: 0.0.1
+spec:
+  configuration:
+    values:
```

Comment:
This is the defaults for the values passed to templates like `cloudformation.stacks.{controlPlane,nodePool,root}.resources.append` and `kubernetes.apiserer.flags[].value`.

The defaults can be overridden from `cluster.yaml` like done in https://github.com/kubernetes-incubator/kube-aws/pull/791/files#diff-941dbd01d3b0e4ed2e77ddcc4e23c43fR52